src/backend/cuda/pool.cu

```diff
@@ -0,0 +1,394 @@
+               const dim_t padding_width, const dim_t padding_height,
+               const dim_t stride_width, const dim_t stride_height,
+               af_pooling_type pool_type)
+{
+    if(supported_cudnn_type<T>())
+        return pool2_cudnn(in, pool_width, pool_height, padding_width, padding_height, stride_width, stride_height, pool_type);
+    else
+        return pool2_native(in, pool_width, pool_height, padding_width, padding_height, stride_width, stride_height, pool_type);
+}
+
+#define INSTANTIATE(T)                                                  \
+    template Array<T> pool2<T>(const Array<T> &in,                      \
+               const dim_t pool_width, const dim_t pool_height,         \
+               const dim_t padding_width, const dim_t padding_height,   \
+               const dim_t stride_width, const dim_t stride_height,     \
+               af_pooling_type pool_type);
+
+INSTANTIATE(float)
+INSTANTIATE(double)
+INSTANTIATE(int)
+INSTANTIATE(uint)
+INSTANTIATE(short)
+INSTANTIATE(ushort)
+INSTANTIATE(uchar)
+
+#undef INSTANTIATE
+
+
+template<typename T>
+Array<T> pool2Grad_native(const Array<T>& incoming_gradient,
+                          const Array<T>& original_input,
+                          const Array<T>& pooled_output,
+                          const dim_t pool_width, const dim_t pool_height,
+                          const dim_t padding_width, const dim_t padding_height,
+                          const int stride_width, const int stride_height,
+                          af_pooling_type pool_type)
+{
+    Array<T> in = incoming_gradient;
+    in.eval();
+
+    Array<T> unwrapped = unwrap(in, pool_width, pool_height,
+                                   stride_width, stride_height,
+                                   padding_width, padding_height, true);
+
+    dim4 udims = unwrapped.dims();
+    udims[0] -= 1;
+    Array<T> max_unwrapped = reduce<af_max_t, T, T>(unwrapped, 0);
+
+
+    dim_t outputWidth  = 1 + (in.dims()[0] + 2 * padding_width - pool_width)   / stride_width;
+    dim_t outputHeight = 1 + (in.dims()[1] + 2 * padding_height - pool_height) / stride_height;
+
+    max_unwrapped.modDims(dim4(outputWidth, outputHeight, in.dims()[2], in.dims()[3]));
+    Array<T> out = createValueArray<T>(max_unwrapped.dims(), 0);
+    out = max_unwrapped;
+
+    return out;
+}
+
+template<typename T>
+Array<T> pool2Grad_cudnn(const Array<T>& incoming_gradient,
+                         const Array<T>& original_input,
+                         const Array<T>& pooled_output,
+                         const dim_t pool_width, const dim_t pool_height,
+                         const dim_t padding_width, const dim_t padding_height,
+                         const int stride_width, const int stride_height,
+                         af_pooling_type pool_type)
+{
+    dim4 idims = incoming_gradient.dims();
+
+    cudnnHandle_t cudnn;
+    CUDNN(cudnnCreate(&cudnn));
```

Comment:
this is should move into platform.hpp/cpp and should be `thread_local` there if cudnn is similar to cublas or other libs from nvidia.

We shouldn't create cudnn context everytime this function is used.

Reply:
Also, this is may be causing a leak since there are no matching `cudnnDestroy` calls.